### PR TITLE
Fatal Error making contribution after redirect Issue #118

### PR DESCRIPTION
**Pull request by aniesshsethh**

### CODE REVIEW

**Comment by magnolia61** [comment]
Am I right if I believe your solution is only a solution for membership payments (and not regular payments or event payments)?

**Comment by aniesshsethh** [comment]
For any payments where there is a thank you page as the code is executed only when CRM_Contribute_Form_Contribution_ThankYou

File: gdpr.php [hunk 1]
```diff
@@ -255,8 +255,7 @@ function gdpr_civicrm_buildForm($formName, $form) {
     }
   }
   if ($formName == 'CRM_Contribute_Form_Contribution_ThankYou') {
-    $trxnId = $form->getVar('_trxnId');
-    $cid    = CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Contribution', $trxnId, 'contact_id', 'trxn_id');
+    $cid = $form->_membershipContactID;
```
Comment (jitendrapurohit) [inline gdpr.php]: I think better to use `$form->_contactID` instead of the above variable. @aniesshsethh Can you verify if this works for you?

Comment (aniesshsethh) [reply gdpr.php]: Yeah, that works too, the reason I didn't used that in order to handle the condition where somebody else is registering for somebody else, I may be wrong there though. Please confirm and I Will change.

Comment (magnolia61) [reply gdpr.php]: $form->_contactID works for us (just tested), but I agree wih @aniesshsethh this should be doublechecked. What if a person pays for someone else with a Pay Now link. The right person should be identified in those scenarios, not just the logged in user.

**Comment by cmolava** [comment]
The means of getting the contact ID for the contribution from the ThankYou form appears to vary. 
$form->_contactID is empty during anonymous submissions. The original approach does work under some cases for anonymous users, but it appears not all cases. Have gone for a combination of both, checking for values to prevent the fatal errors. 

The section of code deals with creating a personalized link to the Communications Preferences form.  In this context, people should only complete their own Communications Preferences, not fill them in on behalf of others. So if there is a logged in user, it is their contact id that is used. (Admins can complete the form for others from the contact action link, in which case the resulting activity is reflected as such).

Thanks for the PR, it prompted a  useful discussion.

**Comment by chriscant** [comment]
Using $cid = $form->_contactID fixed a problem we'd had with all membership payments on CiviCRM 4.7.27, Drupal 7.59 and PayPal Website Payments Pro. We've not had events or other payments through.
It would be good to have this as an official release. Thanks.

**Pull request closed without merging**